### PR TITLE
feat: add realtime voice vision assistant stack

### PR DESCRIPTION
**Pull request by matheus-rech**
## Summary
- scaffold monorepo with Next.js 14 frontend and Express backend for realtime multimodal assistant
- implement backend services for ephemeral key minting, visual frame processing, websocket relay, and session management with tests
- build frontend voice+vision client with realtime session hook, visual capture pipeline, adaptive UI, and supporting utilities/tests

## Testing
- npm run test --workspace backend
- npm run test --workspace frontend
- npm run lint --workspace backend
- npm run lint --workspace frontend

------
https://chatgpt.com/codex/tasks/task_e_68cbbf9f44f883299d2cffb4765f7f60

## Summary by Sourcery

Add a realtime multimodal assistant stack with voice and vision capabilities across a new monorepo setup

New Features:
- Scaffold monorepo with Next.js 14 frontend and Express backend for a realtime voice+vision assistant
- Add frontend useRealtimeSession hook to orchestrate WebSocket connections, audio streaming, and visual frame capture
- Implement VisualProcessor on the frontend for change-detection frame sampling and dynamic cropping
- Build backend services for ephemeral key minting, visual frame processing via sharp over WebSocket, and session management with Redis fallback
- Develop UI components for video preview, transcript display, visual context indicator, mode selector, and connection status

Enhancements:
- Introduce dual-queue interruption management to clear audio buffers and pause visual capture
- Implement ContextManager for sliding-window context maintenance with token-based compression and visual history trimming
- Add ResilientConnection to handle WebSocket reconnection and rate-limit backoff

Documentation:
- Provide a comprehensive README outlining features, monorepo layout, setup, testing, and deployment instructions

Tests:
- Add Vitest test suites for frontend frame utilities and backend services including ContextManager, VisualContextCache, and VisualProcessingService

### CODE REVIEW

**Comment by matheus-rech** [comment]
@codex review

File: apps/frontend/lib/hooks/useRealtimeSession.ts [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+    disconnect,
+    sendText,
+    handleInterrupt,
+    setMode: updateMode,
+    mode,
+    visualActive,
+    latencyMs,
+    previewFrame
+  };
+};
```
[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/frontend/lib/hooks/useRealtimeSession.ts]: @codex review

File: apps/backend/src/services/SessionStore.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      return;
+    }
+    record.lastSeen = Date.now();
+    await this.create(record);
+  }
+
+  async count(): Promise<number> {
+    if (this.redis) {
+      const keys = await this.redis.keys(this.key('*'));
+      return keys.length;
```
[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/backend/src/services/SessionStore.ts:55]: @copilot review

File: apps/backend/src/services/ContextManager.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      }
+      runningTokens = this.entries.reduce((acc, item) => acc + item.tokens, 0);
+    }
+
+    // Hard trim oldest entries if still above limit
+    while (this.entries.reduce((acc, entry) => acc + entry.tokens, 0) > this.options.maxTokens && this.entries.length > 0) {
+      this.entries.shift();
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/backend/src/services/ContextManager.ts:98]: @copilot

File: apps/frontend/lib/hooks/useRealtimeSession.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/frontend/lib/hooks/useRealtimeSession.ts]: @copilot

[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/frontend/lib/hooks/useRealtimeSession.ts]: /claude review

File: apps/backend/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      }
+    }
+  });
+
+  socket.on('message', async (raw) => {
+    try {
+      const data = JSON.parse(raw.toString()) as {
+        type: string;
+        payload?: unknown;
+      };
```
[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/backend/src/index.ts:57]: @sourcery-ai review

File: apps/backend/src/services/ContextManager.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/backend/src/services/ContextManager.ts:98]: @codex address that feedback

File: apps/backend/src/index.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/backend/src/index.ts:57]: @codex address that feedback

File: apps/frontend/lib/hooks/useRealtimeSession.ts [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+      }
+    });
+
+    client.on('error', (error: unknown) => {
+      console.error('Realtime client error', error);
+      setConnectionState('error');
+    });
+  }, []);
+
+  const startVisualPipeline = useCallback(
```
[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/frontend/lib/hooks/useRealtimeSession.ts:135]: @copilot review

File: apps/frontend/lib/hooks/useRealtimeSession.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/frontend/lib/hooks/useRealtimeSession.ts]: @codex address that feedback

File: apps/frontend/lib/audio/AudioController.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  constructor(options: AudioControllerOptions) {
+    this.onChunk = options.onChunk;
+  }
+
+  async start(): Promise<void> {
+    if (this.mediaRecorder) {
+      return;
+    }
+    this.stream = await navigator.mediaDevices.getUserMedia({ audio: true });
+    this.mediaRecorder = new MediaRecorder(this.stream, {
```
[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/frontend/lib/audio/AudioController.ts:30]: @gemini-code-assist please address the comments from this code review:

## Overall Comments
- The useRealtimeSession hook is very large and mixes connection, audio, and visual logic—consider splitting it into smaller hooks (e.g. useVisualPipeline, useAudioCapture, useWebSocketConnection) to improve readability and testability.
- You have hard-coded frame-rate and resolution defaults in both the frontend VisualProcessor and backend env vars—extract these into a shared config or .env so they don’t drift out of sync.
- Both client and server manage WebSocket reconnects and backoff differently—consolidating this into a common resilient-connection utility with consistent retry/backoff behavior would improve reliability.

## Individual Comments

### Comment 1
<location> `apps/frontend/lib/audio/AudioController.ts:30` </location>
<code_context>
+      return;
+    }
+    this.stream = await navigator.mediaDevices.getUserMedia({ audio: true });
+    this.mediaRecorder = new MediaRecorder(this.stream, {
+      mimeType: 'audio/webm;codecs=opus'
+    });
</code_context>

<issue_to_address>
**suggestion:** Consider handling MediaRecorder errors and state transitions.

Currently, errors or unexpected state changes in MediaRecorder are not handled, which could leave the component in an inconsistent state. Adding event listeners for 'error', 'stop', 'pause', and 'resume' would make the implementation more robust.

Suggested implementation:

```typescript
    this.mediaRecorder = new MediaRecorder(this.stream, {
      mimeType: 'audio/webm;codecs=opus'
    });

    this.mediaRecorder.addEventListener('dataavailable', async (event) => {
      if (event.data.size <= 0) {
        return;
      }
      const buffer = await event.data.arrayBuffer();
      const chunk: AudioChunk = { id: crypto.randomUUID(), data: buffer, timestamp: Date.now() };
      this.queue.enqueue(chunk);
      this.onChunk(chunk);
    });

    this.mediaRecorder.addEventListener('error', (event) => {
      console.error('MediaRecorder error:', event);
      // Optionally, clean up or notify the user
      this.cleanup();
    });

    this.mediaRecorder.addEventListener('stop', () => {
      // Handle recorder stop, e.g., cleanup resources
      this.cleanup();
    });

    this.mediaRecorder.addEventListener('pause', () => {
      // Optionally handle pause state
      console.info('MediaRecorder paused');
    });

    this.mediaRecorder.addEventListener('resume', () => {
      // Optionally handle resume state
      console.info('MediaRecorder resumed');
    });

```

You should implement a `cleanup()` method in your class to handle resource cleanup (e.g., stopping tracks, resetting state) when errors or stop events occur. If you want more advanced error handling or user notifications, you can expand the event handlers accordingly.
</issue_to_address>

### Comment 2
<location> `apps/frontend/lib/hooks/useRealtimeSession.ts:239` </location>
<code_context>
+    setPreviewFrame(null);
+  }, [teardownVisualSocket]);
+
+  const sendText = useCallback((text: string) => {
+    if (!clientRef.current) {
+      return;
</code_context>

<issue_to_address>
**issue (bug_risk):** Transcript updates may result in duplicate user messages.

Both the event handler and sendText add user messages to the transcript, which may cause duplicates if the backend also echoes messages. Please implement deduplication or coordinate message addition to prevent this.
</issue_to_address>

### Comment 3
<location> `apps/frontend/lib/hooks/useRealtimeSession.ts:135` </location>
<code_context>
+    });
+  }, []);
+
+  const startVisualPipeline = useCallback(
+    async (nextMode: AssistantMode) => {
+      const processor = visualProcessorRef.current ?? new VisualProcessor({
</code_context>

<issue_to_address>
**issue:** MediaStream acquisition does not handle permission denial or errors.

Catch errors from getUserMedia/getDisplayMedia and update the pipeline state to prevent partial initialization when permissions are denied or acquisition fails.
</issue_to_address>

### Comment 4
<location> `apps/frontend/lib/hooks/useRealtimeSession.ts:268` </location>
<code_context>
+
+  const [message, setMessage] = useState('');
+
+  useEffect(() => {
+    void connect();
+    return () => {
</code_context>

<issue_to_address>
**suggestion (bug_risk):** Disconnect is called on unmount, but may race with connect.

If connect hasn't completed before unmount, disconnect may run prematurely, causing inconsistent state or resource leaks. Consider tracking the connection status or implementing cancellation.

Suggested implementation:

```typescript
  const [connecting, setConnecting] = useState(false);
  const connectAbortControllerRef = useRef<AbortController | null>(null);

  useEffect(() => {
    setConnecting(true);
    const abortController = new AbortController();
    connectAbortControllerRef.current = abortController;

    void (async () => {
      try {
        await connect({ signal: abortController.signal });
        setConnecting(false);
      } catch (err) {
        if (abortController.signal.aborted) {
          // Connection was cancelled
          setConnecting(false);
        }
      }
    })();

    return () => {
      if (connectAbortControllerRef.current) {
        connectAbortControllerRef.current.abort();
      }
      // Only disconnect if we are not in the middle of connecting
      if (!connecting) {
        disconnect();
      }
    };
  }, [connect, disconnect]);

```

1. You will need to update your `connect` function to accept an optional `{ signal }` parameter and handle cancellation using the `AbortController.signal`.
2. If your connection logic does not support cancellation, you may need to refactor it to check for `signal.aborted` and abort early.
3. Make sure to initialize `connecting` state and `connectAbortControllerRef` at the top of your hook.
</issue_to_address>

### Comment 5
<location> `apps/backend/src/services/VisualProcessingService.ts:57` </location>
<code_context>
+    const jpeg = await resized.jpeg({ quality: Math.round(compressionQuality * 100) }).toBuffer();
+    const base64 = jpeg.toString('base64');
+
+    const processed: ProcessedFrame = {
+      base64,
+      width: metadata.width ?? maxResolution,
</code_context>

<issue_to_address>
**issue (bug_risk):** ProcessedFrame width and height may not match resized image.

Use the resized image's dimensions instead of metadata.width and metadata.height to ensure the ProcessedFrame accurately reflects the image size.
</issue_to_address>

### Comment 6
<location> `apps/frontend/tests/frameUtils.spec.ts:3` </location>
<code_context>
+# Realtime Multimodal Assistant
+
+This repository contains a production-ready reference implementation for a realtime voice + vision assistant powered by OpenAI's GPT-Realtime stack. The project delivers a full-stack TypeScript solution with a Next.js 14 frontend and an Express backend that orchestrate audio streaming, frame sampling, and secure ephemeral key generation.
+
</code_context>

<issue_to_address>
**suggestion (testing):** Edge cases for computeFrameDiff and cropToRegion are not fully covered.

Add tests for computeFrameDiff with null previous frames, differing image sizes, and unchanged pixels. For cropToRegion, include cases with null bounds and bounds exceeding canvas size.
</issue_to_address>

### Comment 7
<location> `apps/backend/src/services/RealtimeSessionManager.ts:41-42` </location>
<code_context>
        const payload = (await response.json()) as EphemeralKeyResponse;
        return payload;

</code_context>

<issue_to_address>
**suggestion (code-quality):** Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://docs.sourcery.ai/Reference/Rules-and-In-Line-Suggestions/TypeScript/Default-Rules/inline-immediately-returned-variable))

```suggestion
        return (await response.json()) as EphemeralKeyResponse;

```

<br/><details><summary>Explanation</summary>Something that we often see in people's code is assigning to a result variable
and then immediately returning it.

Returning the result directly shortens the code and removes an unnecessary
variable, reducing the mental load of reading the function.

Where intermediate variables can be useful is if they then get used as a
parameter or a condition, and the name can act like a comment on what the
variable represents. In the case where you're returning it from a function, the
function name is there to tell you what the result is, so the variable name
is unnecessary.
</details>
</issue_to_address>

### Comment 8
<location> `apps/frontend/lib/hooks/useRealtimeSession.ts:64-68` </location>
<code_context>
          const next = [
            { description: payload.description, capturedAt: payload.frame.capturedAt, source: payload.frame.source },
            ...contexts
          ].slice(0, 10);
          return next;

</code_context>

<issue_to_address>
**suggestion (code-quality):** Inline variable that is immediately returned ([`inline-immediately-returned-variable`](https://docs.sourcery.ai/Reference/Rules-and-In-Line-Suggestions/TypeScript/Default-Rules/inline-immediately-returned-variable))

```suggestion
          return [
                      { description: payload.description, capturedAt: payload.frame.capturedAt, source: payload.frame.source },
                      ...contexts
                    ].slice(0, 10);

```

<br/><details><summary>Explanation</summary>Something that we often see in people's code is assigning to a result variable
and then immediately returning it.

Returning the result directly shortens the code and removes an unnecessary
variable, reducing the mental load of reading the function.

Where intermediate variables can be useful is if they then get used as a
parameter or a condition, and the name can act like a comment on what the
variable represents. In the case where you're returning it from a function, the
function name is there to tell you what the result is, so the variable name
is unnecessary.
</details>
</issue_to_address>

### Comment 9
<location> `apps/frontend/lib/hooks/useRealtimeSession.ts:115` </location>
<code_context>
      const item = event.item;

</code_context>

<issue_to_address>
**suggestion (code-quality):** Prefer object destructuring when accessing and using properties. ([`use-object-destructuring`](https://docs.sourcery.ai/Reference/Rules-and-In-Line-Suggestions/TypeScript/Default-Rules/use-object-destructuring))

```suggestion
      const {item} = event;
```

<br/><details><summary>Explanation</summary>Object destructuring can often remove an unnecessary temporary reference, as well as making your code more succinct.

From the [Airbnb Javascript Style Guide](https://airbnb.io/javascript/#destructuring--object)
</details>
</issue_to_address>

### Comment 10
<location> `apps/frontend/lib/visual/VisualProcessor.ts:162` </location>
<code_context>
        const result = reader.result;

</code_context>

<issue_to_address>
**suggestion (code-quality):** Prefer object destructuring when accessing and using properties. ([`use-object-destructuring`](https://docs.sourcery.ai/Reference/Rules-and-In-Line-Suggestions/TypeScript/Default-Rules/use-object-destructuring))

```suggestion
        const {result} = reader;
```

<br/><details><summary>Explanation</summary>Object destructuring can often remove an unnecessary temporary reference, as well as making your code more succinct.

From the [Airbnb Javascript Style Guide](https://airbnb.io/javascript/#destructuring--object)
</details>
</issue_to_address>

[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/frontend/lib/audio/AudioController.ts:30]: Here's the updated code for the start() method:

Suggested change
    this.mediaRecorder = new MediaRecorder(this.stream, {
    this.mediaRecorder = new MediaRecorder(this.stream, {
      mimeType: 'audio/webm;codecs=opus'
    });
    this.mediaRecorder.addEventListener('dataavailable', async (event) => {
      if (event.data.size <= 0) {
        return;
      }
      const buffer = await event.data.arrayBuffer();
      const chunk: AudioChunk = { id: crypto.randomUUID(), data: buffer, timestamp: Date.now() };
      this.queue.enqueue(chunk);
      this.onChunk(chunk);
    });
    this.mediaRecorder.addEventListener('error', (event) => {
      console.error('MediaRecorder error:', event);
      this.stop(); // Reuse existing stop method for cleanup
    });
    this.mediaRecorder.addEventListener('stop', () => {
      this.stop(); // Reuse existing stop method for cleanup
    });
    this.mediaRecorder.addEventListener('pause', () => {
      console.info('MediaRecorder paused');
    });
    this.mediaRecorder.addEventListener('resume', () => {
      console.info('MediaRecorder resumed');
    });
    this.mediaRecorder.start(200);

File: apps/frontend/lib/hooks/useRealtimeSession.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/frontend/lib/hooks/useRealtimeSession.ts]: @codex address that feedback

File: apps/backend/src/services/ContextManager.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/backend/src/services/ContextManager.ts:98]: @copilot address this

**Review by matheus-rech — COMMENTED** [review]
ok

File: apps/backend/src/services/VisualProcessingService.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+  async describeFrame(frame: ProcessedFrame, quick = false): Promise<string> {
+    const cached = this.cache.get(frame.hash);
+    if (cached) {
+      return cached.description;
+    }
+
+    // TODO: Replace this placeholder implementation with actual OpenAI Vision API calls.
+    // This is a temporary solution for development/offline mode.
+    // Consider adding a configuration flag to control placeholder vs. real API usage.
```
[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/backend/src/services/VisualProcessingService.ts:78]: @copilot address it now

[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/backend/src/services/VisualProcessingService.ts:78]: @gemini-code-assist address this issue

**Comment by matheus-rech** [comment]
This pull request introduces the initial backend implementation for a realtime multimodal assistant, establishing a robust TypeScript/Express server with support for audio and visual streaming, session management, context handling, and secure ephemeral key generation. The codebase is structured for scalability and maintainability, including essential middleware, WebSocket handling, Redis-backed session storage, and comprehensive ESLint/test configurations.

Key backend features and structure:

**Core Backend Architecture**

- Introduces the main Express server setup in `src/app.ts` and `src/index.ts`, integrating security, compression, CORS, rate limiting, and error handling middleware. WebSocket support is provided for realtime visual data relaying, and the backend orchestrates context management, visual processing, and dual-queue interruption handling. [[1]](diffhunk://#diff-56c5b26f47d61c37f4bdc9bb5ae17525c6d31614ce3633b4c3b5047303859a25R1-R34) [[2]](diffhunk://#diff-9f9b2e661dbb7c922ab7db42bebcf7f1e1d07de29477a7c6265daaa2fa6dc5eaR1-R102)

**Session and Key Management**

- Implements secure ephemeral session key minting via the OpenAI Realtime API, with retry logic, and manages session records using Redis (with in-memory fallback). Session limits and context are enforced, and endpoints are provided for session creation and health checks. [[1]](diffhunk://#diff-44be55e78fed2fad6ebf62e11988443e4d19aa8f5a3eaf36aa353fdf5dbeb6d7R1-R54) [[2]](diffhunk://#diff-f94c82ea11476d165078abf38de30796ac6f5212d280b3b97a057fff1762f941R1-R63) [[3]](diffhunk://#diff-38dd1306279e541301909817a6f8016b37d737b7a895de3efe3fcdc5c8b50c1eR1-R32) [[4]](diffhunk://#diff-3236462973dd48a636e88ea2023aec3dace542c83993d358a3fe36fbdebd9ca9R1-R5)

**Context and Visual Processing**

- Adds a `ContextManager` for maintaining and compressing sliding windows of audio/visual/system context, and a `VisualContextCache` for deduplication and retrieval of visual memory. Includes dual-queue logic for synchronized audio/visual interruption and resumption. [[1]](diffhunk://#diff-e96a7186300ae0a51ce752d5a9c73d50359db622cdeddddd2e217f32124e5ce6R1-R98) [[2]](diffhunk://#diff-3922482ab9ba8b7328688e5d783040886a8920ce92d042aacbda6620c4de43ecR1-R79)

**Testing and Quality**

- Provides initial unit tests for context management and visual caching, and sets up a comprehensive ESLint configuration (with TypeScript, import, and security plugins) and test scripts using Vitest. [[1]](diffhunk://#diff-1fec33c286532485e1f985706e6e5154f4e51eb66b78942e51397881dd49204fR1-R19) [[2]](diffhunk://#diff-f54a7ce9ec060cf66ec7533ad68280d4e09cda25746844e876055f6162e9b210R1-R13) [[3]](diffhunk://#diff-3a7076676cba00309eaa64fe6ab63020bbc8d62ec5021fcdbf13db86f4d99ba7R1-R35) [[4]](diffhunk://#diff-8d3b741039fe624dded05aa46e994d93ec459ac8d8a0cfc747f92ef447a9513aR1-R47)

**Documentation**

- Adds a detailed `README.md` outlining features, monorepo structure, setup steps, testing, and deployment considerations.

These changes lay the foundation for a scalable, secure, and feature-rich backend for the multimodal assistant project.

---

**Backend Architecture and API**

- Establishes Express server with security, compression, CORS, rate limiting, error handling, and JSON body parsing; sets up WebSocket for realtime visual data relay. (`src/app.ts`, `src/index.ts`) [[1]](diffhunk://#diff-56c5b26f47d61c37f4bdc9bb5ae17525c6d31614ce3633b4c3b5047303859a25R1-R34) [[2]](diffhunk://#diff-9f9b2e661dbb7c922ab7db42bebcf7f1e1d07de29477a7c6265daaa2fa6dc5eaR1-R102)
- Implements health check and session creation API endpoints, including session limit enforcement and error handling. (`src/routes/health.ts`, `src/routes/session.ts`) [[1]](diffhunk://#diff-3236462973dd48a636e88ea2023aec3dace542c83993d358a3fe36fbdebd9ca9R1-R5) [[2]](diffhunk://#diff-38dd1306279e541301909817a6f8016b37d737b7a895de3efe3fcdc5c8b50c1eR1-R32)

**Session and Key Management**

- Implements ephemeral session key minting with OpenAI Realtime API and retry/backoff logic. (`src/services/RealtimeSessionManager.ts`)
- Provides Redis-backed session storage with fallback to in-memory store; supports session creation, retrieval, update, and counting. (`src/services/SessionStore.ts`)

**Context and Visual Processing**

- Adds `ContextManager` for sliding window context with compression and visual history limits. (`src/services/ContextManager.ts`)
- Introduces dual-queue structure for audio and visual stream management, supporting interruption and resumption. (`src/queues/DualQueue.ts`)
- Adds resilient WebSocket connection handling with backoff and rate limiting. (`src/services/ResilientConnection.ts`)

**Testing and Quality**

- Sets up unit tests for context management and visual cache. (`src/services/ContextManager.spec.ts`, `src/services/VisualContextCache.spec.ts`) [[1]](diffhunk://#diff-1fec33c286532485e1f985706e6e5154f4e51eb66b78942e51397881dd49204fR1-R19) [[2]](diffhunk://#diff-f54a7ce9ec060cf66ec7533ad68280d4e09cda25746844e876055f6162e9b210R1-R13)
- Adds ESLint configuration for code quality and security, and updates `package.json` with scripts and dependencies. (`.eslintrc.cjs`, `package.json`) [[1]](diffhunk://#diff-3a7076676cba00309eaa64fe6ab63020bbc8d62ec5021fcdbf13db86f4d99ba7R1-R35) [[2]](diffhunk://#diff-8d3b741039fe624dded05aa46e994d93ec459ac8d8a0cfc747f92ef447a9513aR1-R47)

**Documentation**

- Adds a comprehensive `README.md` describing features, structure, setup, testing, and deployment. (`README.md`)

File: apps/backend/src/services/VisualProcessingService.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (matheus-rech) [reply apps/backend/src/services/VisualProcessingService.ts:78]: @copilot

**Review by matheus-rech — COMMENTED** [review]
agree
